# **Tarea 1 - CC6205 Natural Language Processing 📚**

**Integrantes:**

**Fecha límite de entrega 📆:** Lunes 05 de Abril.

**Tiempo estimado de dedicación:**




` ` 


Bienvenid@s a la primera tarea en el curso de Natural Language Processing (NLP). Esta tarea tiene como objetivo evaluar los contenidos teóricos de las primeras semanas de clases, enfocado principalmente en ***Information Retrieval (IR)*** y ***Vector Space Models***. Si aún no han visto las clases, se recomienda visitar los links de las referencias.

La tarea consta de una parte teórica que busca evaluar conceptos vistos en clases. Seguido por una parte práctica con el fín de introducirlos a la programación en Python enfocada en NLP. 

` ` 




**Instrucciones:**
- La tarea se realiza en grupos de **máximo** 2 personas. Puede ser invidivual pero no es recomendable.
- La entrega es a través de u-cursos a más tardar el día estipulado arriba. No se aceptan atrasos.
- El formato de entrega es este mismo Jupyter Notebook.
- Al momento de la revisión tu código será ejecutado. Por favor verifica que tu entrega no tenga errores de compilación. 
- Está **PROHIBIDO** usar cualquier librería que implemente los algoritmos pedidos (Spacy, scikit, etc). Sólo se podrán utilizar las librerías importadas al inicio de la sección de práctica.
- En el horario de auxiliar pueden realizar consultas acerca de la tarea a través del canal de Discord del curso. 



Ahora sí, empecemos! 😄🚀

` ` 

**Referencias:**


Slides:
    
- [Introducción al curso](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-introduction.pdf)
- [Vector Space Model / Information Retrieval](https://github.com/dccuchile/CC6205/blob/master/slides/NLP-IR.pdf)    

Videos: 

- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte I](https://www.youtube.com/watch?v=HEKTNOttGvU)
- [CC6205 - Procesamiento de Lenguaje Natural: CC6205 - Procesamiento de Lenguaje Natural: Introducción parte II](https://www.youtube.com/watch?v=P8cwnI-f-Kg)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 1](https://youtu.be/FXIVClF370w)
- [CC6205 - Procesamiento de Lenguaje Natural: Vector Space Model and Information Retrieval parte 2](https://youtu.be/f8nG1EMmPZk)

` ` 



## **1 - Preguntas teóricas 📕 (2 puntos).** ##









Las siguientes celdas contienen preguntas acerca del contenido visto en clases y en el material del curso.  Contestar cada pregunta en su celda correspondiente y **no extenderse más de 5 lineas** . 🙏
` ` 

**Pregunta 1 (0.5 puntos): ¿Por qué el análisis del lenguaje humano es una tarea compleja? Mencione dos razones.**



Porque es extremadamente complejo entender formalmente y describir las reglas que rigen el lenguaje. El lenguaje humano es muy ambiguo y es dinámico.



**Pregunta 2 (0.5 puntos): ¿Cuál es la diferencia entre Natural Language Processing (NLP) y Computational Linguistics?**

NLP desarrolla métodos y modelos para resolver problemas prácticos (tasks) del lenguaje. Por otro lado, CL es una disciplina que a través de métodos estadísticos y computacionales busca entender procesos y formalismos del lenguaje humano.


**Pregunta 3 (0.5 puntos): ¿Qué es el Machine Learning supersivado? ¿Por qué es importante para el campo de NLP?**

Corresponde a un paradigma de aprendizaje en el área de Machine Learning. Se basa en la creación de algoritmos capaces de reconocer las relaciones entre variables de entradas y de salida a partir de un conjunto de datos etiquetados previamente. Es muy útil en el área de NLP por la dificultad de formalizar el lenguaje, entregando excelentes resultados en la mayoría de las tasks existentes.

**Pregunta 4 (0.5 puntos): ¿Cuáles son las diferencias entre usar Deep learning y Machine Learning clásico (empirismo) para un problema de NLP? Ejemplifique con alguna task.**

En machine learning clásico el desarrollador debía crear por el mismo las características/representaciones. En deep learning, estas representaciones son aprendidas por los mismos algoritmos.

## **2 - Preguntas prácticas 💻 (4 puntos).** ##

Esta segunda sección incluye ejercicios de programación 🤙. Leer atentamente las instrucciones entregadas a continuación para facilitar el proceso de revisión de sus trabajos.


` ` 

**Instrucciones:**



- Escribe tu código entre las lineas de comentarios **### Aquí inicia tu código ###** y **### Aquí termina tu código ###**.
- Cuando el ejercicio incluya un bloque llamado ***Test***, comprueba que el resultado de la ejecución coincida con el resultado esperado.
- Recuerde siempre mantener buenas prácticas de código.
- Está permitido sólo utilizar las librerías importadas antes del Ejercicio 1.
- **Recordar** que: *Documento = Oración. Dataset = Corpus. Vocabulario = Typos*.




**Ejemplo:** Implemente una función **`hello_world()`** que imprima en pantalla `"Hello World"`. 

In [ ]:
def hello_world():
  ### Aquí inicia tu código ###
  print("Hello World")
  ### Aquí termina tu código ###

***Test:***

In [ ]:
hello_world()

Hello World


***Resultado esperado***: 
<table>
    <tr> 
        <td> Hello World </td> 
    </tr>
</table> 

``
``




Estas son las librerías permitidas. Si quieren utilizar alguna librería adicional, pueden realizar la consulta a través de Discord. 

In [ ]:
import codecs
import re
import numpy as np
import pandas as pd

` ` 

**Ejercicio 1 - *Tokenización* (0.5 puntos).** 
` `  
` ` 

En el primer ejercicio veremos la dificultad 😨 de tokenizar textos no estructurados, destacando la importancia de tener librerías que realicen este trabajo. 

El archivo adjunto al enunciado de la tarea contiene una interconsulta médica real en Chile, estos textos son frecuentemente utilizados para obtener información epidemiológica y realizar estudios estadísticos a partir de ellos. Es por esto que poder analizar el texto de manera correcta es fundamental. 


Ejecute el código a continuación para cargar el ejemplo. Recuerde realizar la modificación al directorio en caso que el archivo no se encuentre en el mismo directorio del Jupyter Notebook.

In [ ]:
text = codecs.open('interconsulta.txt', 'r', 'UTF-8').read()
print(text)

Interconsulta n°2 del mes.-05/10/2019
- ENFERMEDAD RENAL  CRÓNICA ETAPA 3. 
- Fundamento Clínico APS: PACIENTE  DE 61 AÑOS, 	CON ANTECEDENTES DE HRTA DE LARGA DATA.ACUDE A CONTROL PSCV CON EXAMNES, DONDE DESTACA CREATININA DE 1,43,
 CON UNA VFG CALCULADA DE 53 ML/MIN/1.73M2. CON UNA CREATININA HACE MAS DE 3 MESES DE 1.34.
                   
Enfermedad renal cronica Etapa 3(SOLICITO EVALUACION DEL CASO).



Implementen una función **`get_tokens()`** que reciba un texto y entregue una lista con sus tokens. Son libres de elegir la forma de tokenizar mientras no utilicen librerías con tokenizadores ya implementados. Pueden utilizar la librería **re** importada para trabajar símbolos.


Ejemplo de uso:

`get_tokens('Este es un ejemplo de prueba.')` 

Nos entregaría:

`['Este', 'es', 'un', 'ejemplo', 'de', 'prueba', '.']`

In [ ]:
def get_tokens(text):
  ### Inicio del código ###
  tokens = []
  tokenizer = re.compile(r'([^\W_]+|.)')
  for line in text.splitlines():
    tokens += [t for t in tokenizer.split(line) if t.strip()!='']
  return tokens
  ### Fín del código ###

In [ ]:
tokens = get_tokens(text)
tokens

['Interconsulta',
 'n',
 '°',
 '2',
 'del',
 'mes',
 '.',
 '-',
 '05',
 '/',
 '10',
 '/',
 '2019',
 '-',
 'ENFERMEDAD',
 'RENAL',
 'CRÓNICA',
 'ETAPA',
 '3',
 '.',
 '-',
 'Fundamento',
 'Clínico',
 'APS',
 ':',
 'PACIENTE',
 'DE',
 '61',
 'AÑOS',
 ',',
 'CON',
 'ANTECEDENTES',
 'DE',
 'HRTA',
 'DE',
 'LARGA',
 'DATA',
 '.',
 'ACUDE',
 'A',
 'CONTROL',
 'PSCV',
 'CON',
 'EXAMNES',
 ',',
 'DONDE',
 'DESTACA',
 'CREATININA',
 'DE',
 '1',
 ',',
 '43',
 ',',
 'CON',
 'UNA',
 'VFG',
 'CALCULADA',
 'DE',
 '53',
 'ML',
 '/',
 'MIN',
 '/',
 '1',
 '.',
 '73M2',
 '.',
 'CON',
 'UNA',
 'CREATININA',
 'HACE',
 'MAS',
 'DE',
 '3',
 'MESES',
 'DE',
 '1',
 '.',
 '34',
 '.',
 'Enfermedad',
 'renal',
 'cronica',
 'Etapa',
 '3',
 '(',
 'SOLICITO',
 'EVALUACION',
 'DEL',
 'CASO',
 ')',
 '.']

**Describa cuáles fueron sus supuestos para realizar la tokenización y compare sus tokens con los entregados por la librería nltk en el bloque de código de más abajo.**

Se utilizaron patrones de expresiones regulares para poder separar correctamente los símbolos en el texto. En este caso, el tokenizador actúa mejor que el de Nltk.

In [ ]:
from nltk.tokenize import wordpunct_tokenize 
nltk_tokens = wordpunct_tokenize(text)
nltk_tokens

NameError: ignored

` `  
` ` 

**Ejercicio 2 - *Stopwords y Stemming* (1 punto).** 
` `  
` ` 

Considere el siguiente corpus:


In [ ]:
dataset = ["I like human languages", "I like programming languages", "Spanish is my favorite language"]

Diseñe una función **`get_vocab()`** que extraiga los typos de este corpus solamente tokenizando. Puede utilizar la función del Ejercicio 1.

In [ ]:
def get_vocab(dataset):
  ### Aquí inicia tu código ###
  tokens = []
  for sent in dataset:
    tokens += get_tokens(sent)
  return list(set(tokens))
  ### Aquí termina tu código ###

***Test:***

In [ ]:
vocab = get_vocab(dataset)
vocab

['favorite',
 'language',
 'Spanish',
 'is',
 'my',
 'I',
 'languages',
 'human',
 'programming',
 'like']

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td>['favorite',
 'Spanish',
 'language',
 'I',
 'like',
 'programming',
 'languages',
 'my',
 'human',
 'is'] </td> 
    </tr>
</table> 

``
``

Ahora diseñe reglas que usted estime convenientes tanto de **Stemming** como de **Stopwords**. Implemente una función que reciba una lista con los elementos del vocabulario, le aplique sus reglas y devuelva el vocabulario actualizado. 

Stopwords: I, is, my
 
Reglas de Stemming: 
 
- ...s -> nada
- ...ing -> nada

In [ ]:
def pre_processing(vocab): 
  ### Aquí inicia tu código ###
  new_vocab = []
  stopwords = ['I', 'is', 'my']
  for word in vocab:
    if word not in stopwords:
      if word.endswith('s'):
        new_vocab.append(word[:-1])
      elif word.endswith('ing'):
        new_vocab.append(word[:-3])
      else:
        new_vocab.append(word)
  return list(set(new_vocab))
  ### Aquí termina tu código ###

In [ ]:
vocab = pre_processing(vocab)
vocab

['favorite', 'language', 'Spanish', 'programm', 'human', 'like']

` `  
` ` 

**Ejercicio 3 - *Bag of Words* 🐶🐈(0.5 puntos).** 
` `  
` ` 

Considere el siguiente corpus, donde cada elemento del arreglo representa un documento:

In [ ]:
d0 = 'El perro se come la comida y después se duerme'
d1 = 'El perro se despierta y después empieza a ladrar'
d2 = 'El perro ladra y después se come la comida'
d3 = 'El gato se come la comida y después se duerme'
d4 = 'El gato se despierta y después empieza a maullar'
d5 = 'El gato maulla y después se come la comida'
dataset = [d0, d1, d2, d3, d4, d5]

El objetivo de este ejercicio es determinar cuáles de  los documentos entregados son los más similares entre sí. Para ello utilizaremos la técnica TF-IDF. 

Como los algoritmos de Machine Learning no comprenden el texto en lenguaje natural, estos documentos deben ser convertidos a vectores numéricos. La representación más simple vista en clases es el **Bag of Words**, método mediante el cuál se cuentan las apariciones de cada palabra en cada uno de los documentos entregados.

Implemente la función **`bag_of_words()`**, que reciba como input un arreglo de documentos y devuelva un pandas dataframe con la representación Bag of Words de los documentos entregados. En esta representación las columnas son el vocabulario y las filas representa las apariciones de cada una de las palabras en los documentos. En otras palabras, cada fila representa el bow de un determinado documento.


Por ejemplo para el siguiente dataset: 

```
dataset = ['El perro ladra', 'El perro come']
```

Debiese entregarnos lo siguiente:


|   | el | perro | ladra | come |
|---|----|-------|------|-------|
| 0 | 1  | 1     | 1    | 0     |
| 1 | 1  | 1     | 0    | 1     |



In [ ]:
def bag_of_words(dataset):
  ### Aquí inicia tu código ###
  all_bow = []
  for document in dataset:
      current_bow = {word: 0 for word in get_vocab(dataset)}
      for word in get_tokens(document):
          current_bow[word] += 1
      all_bow.append(current_bow)
  return pd.DataFrame(all_bow, index = ['d0', 'd1', 'd2', 'd3', 'd4', 'd5'])
  ### Aquí termina tu código ###

***Test:***

In [ ]:
dataset_bow = bag_of_words(dataset)
dataset_bow

maulla  perro  gato  despierta  a  ...  se  empieza  comida  come  duerme
d0       0      1     0          0  0  ...   2        0       1     1       1
d1       0      1     0          1  1  ...   1        1       0     0       0
d2       0      1     0          0  0  ...   1        0       1     1       0
d3       0      0     1          0  0  ...   2        0       1     1       1
d4       0      0     1          1  1  ...   1        1       0     0       0
d5       1      0     1          0  0  ...   1        0       1     1       0

[6 rows x 17 columns]

``
``

***Resultado esperado***: 

|    | El | perro | se | come | la | comida | y | después | duerme | despierta | empieza | a | ladrar | ladra | gato | maullar | maulla |
|----|---:|------:|---:|-----:|---:|-------:|--:|--------:|-------:|----------:|--------:|--:|-------:|------:|-----:|--------:|-------:|
| d0 |  1 |     1 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    0 |       0 |      0 |
| d1 |  1 |     1 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      1 |     0 |    0 |       0 |      0 |
| d2 |  1 |     1 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     1 |    0 |       0 |      0 |
| d3 |  1 |     0 |  2 |    1 |  1 |      1 | 1 |       1 |      1 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      0 |
| d4 |  1 |     0 |  1 |    0 |  0 |      0 | 1 |       1 |      0 |         1 |       1 | 1 |      0 |     0 |    1 |       1 |      0 |
| d5 |  1 |     0 |  1 |    1 |  1 |      1 | 1 |       1 |      0 |         0 |       0 | 0 |      0 |     0 |    1 |       0 |      1 |

``
``

` `  
` ` 

**Ejercicio 4 - *Calcular TF* (0.5 puntos):** Ahora debemos usar el dataframe del ejercicio anterior para calcular la matriz de TF normalizada por la máxima frecuencia ${max_i({\text{tf}_{i,j})}}$, donde
i corresponde al índice de las filas (bow) y j al de las columnas (palabras). Es decir, dividir cada bow en la cantidad de veces de la palabra que aparezca más veces en ese vector (documento). 


$$\text{nft}_{i,j} = \frac{\text{tf}_{i,j}}{max_i({\text{tf}_{i,j})}}$$

In [ ]:
def calc_tf(dataset_bow):
    ### Aquí inicia tu código ###
    return dataset_bow.divide(dataset_bow.max(axis=1), axis=0)
    ### Aquí termina tu código ###

***Test:***

In [ ]:
tf = calc_tf(dataset_bow)
tf

maulla  perro  gato  despierta    a  ...   se  empieza  comida  come  duerme
d0     0.0    0.5   0.0        0.0  0.0  ...  1.0      0.0     0.5   0.5     0.5
d1     0.0    1.0   0.0        1.0  1.0  ...  1.0      1.0     0.0   0.0     0.0
d2     0.0    1.0   0.0        0.0  0.0  ...  1.0      0.0     1.0   1.0     0.0
d3     0.0    0.0   0.5        0.0  0.0  ...  1.0      0.0     0.5   0.5     0.5
d4     0.0    0.0   1.0        1.0  1.0  ...  1.0      1.0     0.0   0.0     0.0
d5     1.0    0.0   1.0        0.0  0.0  ...  1.0      0.0     1.0   1.0     0.0

[6 rows x 17 columns]

``
``

***Resultado esperado***: 

|    |  El | perro |  se | come |  la | comida |   y | después | duerme | despierta | empieza |   a | ladrar | ladra | gato | maullar | maulla |
|----|----:|------:|----:|-----:|----:|-------:|----:|--------:|-------:|----------:|--------:|----:|-------:|------:|-----:|--------:|-------:|
| d0 | 0.5 |   0.5 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d1 | 1.0 |   1.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    1.0 |   0.0 |  0.0 |     0.0 |    0.0 |
| d2 | 1.0 |   1.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   1.0 |  0.0 |     0.0 |    0.0 |
| d3 | 0.5 |   0.0 | 1.0 |  0.5 | 0.5 |    0.5 | 0.5 |     0.5 |    0.5 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  0.5 |     0.0 |    0.0 |
| d4 | 1.0 |   0.0 | 1.0 |  0.0 | 0.0 |    0.0 | 1.0 |     1.0 |    0.0 |       1.0 |     1.0 | 1.0 |    0.0 |   0.0 |  1.0 |     1.0 |    0.0 |
| d5 | 1.0 |   0.0 | 1.0 |  1.0 | 1.0 |    1.0 | 1.0 |     1.0 |    0.0 |       0.0 |     0.0 | 0.0 |    0.0 |   0.0 |  1.0 |     0.0 |    1.0 |

``
``

` `  
` ` 

**Ejercicio 5 - *Calcular IDF* (0.5 punto):**


Implementar `calc_idf(dataset_bow)`. Este debe retornar un diccionario en donde las llaves sean las palabras y los valores sean el calculo de cada idf por palabra.

Recordar que $idf_{t_i} = log_{10}\frac{N}{n_i}$ con $N = $ número de documentos y $n_i = $ Número de documentos que contienen la palabra $t_i$ 

In [ ]:
def calc_idf(dataset_bow):
    ### Aquí inicia tu código ###
    number_of_docs = len(dataset_bow)
    idf = {}
    for word in dataset_bow:
        n_i = (dataset_bow[[word]] != 0).sum()[0]
        idf[word] = np.log10(number_of_docs/n_i)
    return idf
    ### Aquí termina tu código ###

***Test:***

In [ ]:
idf = calc_idf(dataset_bow)
idf

{'El': 0.0,
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0}

``
``

***Resultado esperado***: 
<table>
    <tr> 
        <td> {'El': 0.0,
 'a': 0.47712125471966244,
 'come': 0.17609125905568124,
 'comida': 0.17609125905568124,
 'despierta': 0.47712125471966244,
 'después': 0.0,
 'duerme': 0.47712125471966244,
 'empieza': 0.47712125471966244,
 'gato': 0.3010299956639812,
 'la': 0.17609125905568124,
 'ladra': 0.7781512503836436,
 'ladrar': 0.7781512503836436,
 'maulla': 0.7781512503836436,
 'maullar': 0.7781512503836436,
 'perro': 0.3010299956639812,
 'se': 0.0,
 'y': 0.0} </td> 
    </tr>
</table> 

``
``

Puede notar el bajo puntaje otorgado a las palabras que más se repiten! 😮

**Ejercicio 6 - *Calcular TF-IDF & concluir similitud de documentos.* (1 punto)**


Programe la función `calc_tf_idf(tf, idf)` que entrega el dataframe TF-IDF asociado al dataset que estamos analizando.

In [ ]:
def calc_tf_idf(tf, idf):
    ### Aquí inicia tu código ###
    return tf * idf
    ### Aquí termina tu código ### 

***Test.***

In [ ]:
tf_idf = calc_tf_idf(tf, idf)
tf_idf

maulla     perro      gato  ...    comida      come    duerme
d0  0.000000  0.150515  0.000000  ...  0.088046  0.088046  0.238561
d1  0.000000  0.301030  0.000000  ...  0.000000  0.000000  0.000000
d2  0.000000  0.301030  0.000000  ...  0.176091  0.176091  0.000000
d3  0.000000  0.000000  0.150515  ...  0.088046  0.088046  0.238561
d4  0.000000  0.000000  0.301030  ...  0.000000  0.000000  0.000000
d5  0.778151  0.000000  0.301030  ...  0.176091  0.176091  0.000000

[6 rows x 17 columns]

``
``

***Resultado esperado***: 

|    |  El |    perro |  se |     come |       la |   comida |   y | después |   duerme | despierta |  empieza |        a |   ladrar |    ladra |     gato |  maullar |   maulla |
|----|----:|---------:|----:|---------:|---------:|---------:|----:|--------:|---------:|----------:|---------:|---------:|---------:|---------:|---------:|---------:|---------:|
| d0 | 0.0 | 0.150515 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d1 | 0.0 | 0.301030 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.778151 | 0.000000 | 0.000000 | 0.000000 | 0.000000 |
| d2 | 0.0 | 0.301030 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.778151 | 0.000000 | 0.000000 | 0.000000 |
| d3 | 0.0 | 0.000000 | 0.0 | 0.088046 | 0.088046 | 0.088046 | 0.0 |     0.0 | 0.238561 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.150515 | 0.000000 | 0.000000 |
| d4 | 0.0 | 0.000000 | 0.0 | 0.000000 | 0.000000 | 0.000000 | 0.0 |     0.0 | 0.000000 |  0.477121 | 0.477121 | 0.477121 | 0.000000 | 0.000000 | 0.301030 | 0.778151 | 0.000000 |
| d5 | 0.0 | 0.000000 | 0.0 | 0.176091 | 0.176091 | 0.176091 | 0.0 |     0.0 | 0.000000 |  0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.000000 | 0.301030 | 0.000000 | 0.778151 |


``
``

Ahora que tenemos el dataframe de TF-IDF, nos queda calcular la similitud coseno entre todos los vectores. Notar que la matriz resultante será una matriz simétrica. Implemente la función *cosine_similarity(v1, v2)* que recibe dos vectores (v1 y v2) y calcula la similitud coseno entre ambos vectores. Concluya cuáles son los dos documentos más similares.

In [ ]:
def cosine_similarity(v1, v2):
  ### Aquí inicia tu código ###
  return np.dot(v1,v2)/(np.linalg.norm(v1)*np.linalg.norm(v2))
  ### Aquí termina tu código ### 

similarity_matrix = np.zeros((6,6))
for i, v1 in enumerate(tf_idf.index.values):
  for j, v2 in enumerate(tf_idf.index.values):
      similarity = cosine_similarity(tf_idf.loc[v1].values, tf_idf.loc[v2].values)
      similarity_matrix[i][j] = similarity
      
print(similarity_matrix)

[[1.         0.12032418 0.3223436  0.77967014 0.         0.1632828 ]
 [0.12032418 1.         0.08686457 0.         0.4952126  0.        ]
 [0.3223436  0.08686457 1.         0.1632828  0.         0.11787732]
 [0.77967014 0.         0.1632828  1.         0.12032418 0.3223436 ]
 [0.         0.4952126  0.         0.12032418 1.         0.08686457]
 [0.1632828  0.         0.11787732 0.3223436  0.08686457 1.        ]]


Como era de esperar, con la técnica TF-IDF los documentos más similares son los documentos d0 y d3, ya que difieren en una sola palabra entre muchas. Ahora bien, sabemos que esas dos palabras son muy diferentes entre sí (perro y gato) pero este algoritmo no puede capturar dicha interpretación ya que se basa en técnicas de frecuencias y no de contexto, para ello utilizaremos pronto los Word Embeddings!

![gato](https://www.elagoradiario.com/wp-content/uploads/2020/05/Gato-mascarilla.jpg)

**Cualquier recomendación que nos quisieran dar para una futura tarea es bienvenid@!**